File: backend/models/route.model.js
const mongoose = require('mongoose');

// ─────────────────────────────────────────────────────────────────────────────
// Route model — Piko trails (curated + user-generated).
//
// GeoJSON `geometry` (LineString) drives the map/elevation; `startPoint` (Point,
// 2dsphere) powers maps + "near me". Engagement (savedBy, votes, comments) is
// GLOBAL popularity — the group-scoped decision lives on the Dashboard. A
// `toClient(userId)` projection hides raw id arrays and returns the viewer's own
// saved/vote state, matching the Piko frontend `Route` shape.
// ─────────────────────────────────────────────────────────────────────────────

const commentSchema = new mongoose.Schema(
  {
    id: { type: String, required: true },
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    author: { type: String, default: 'Anonymous' },
    content: { type: String, required: true, maxlength: 2000 },
    createdAt: { type: Date, default: Date.now },
  },
  { _id: false }
);

const voteSchema = new mongoose.Schema(
  {
    user: { type: mongoose.Schema.Types.ObjectId, ref: 'User', required: true },
    value: { type: Number, enum: [-1, 1], required: true },
  },
  { _id: false }
);

const routeSchema = new mongoose.Schema(
  {
    // Stable human id for curated routes (e.g. "morskie-oko"); UGC uses _id.
    slug: { type: String, index: true, unique: true, sparse: true },
    type: { type: String, enum: ['curated', 'user_generated'], default: 'user_generated' },
    title: { type: String, required: true, maxlength: 160 },
    country: { type: String, default: '' },
    location: { type: String, default: '' },
    description: { type: String, default: '' },
    difficulty: { type: String, enum: ['easy', 'moderate', 'hard'], default: 'moderate' },
    distanceKm: { type: Number, default: 0 },
    elevationGainM: { type: Number, default: 0 },
    estimatedDurationMins: { type: Number, default: 0 },
    photos: { type: [String], default: [] },
    tags: { type: [String], default: [] },
    ecoScore: { type: Number, default: 75 },
    weatherScore: { type: Number, default: 70 },
    accessibilityScore: { type: Number, default: 50 },
    ecoImpact: {
      transportMode: { type: String, default: '' },
      co2EstimateKg: { type: Number, default: 0 },
      greenerAlternatives: { type: [String], default: [] },
    },
    geometry: {
      type: { type: String, enum: ['LineString'], default: 'LineString' },
      coordinates: { type: [[Number]], default: undefined }, // [[lng, lat, ele?], …]
    },
    startPoint: {
      type: { type: String, enum: ['Point'] },
      coordinates: { type: [Number] }, // [lng, lat]
    },
    creator: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
    creatorName: { type: String, default: 'Dayla' },
    savedBy: [{ type: mongoose.Schema.Types.ObjectId, ref: 'User' }],
    votes: { type: [voteSchema], default: [] },
    comments: { type: [commentSchema], default: [] },
    moderation: {
      status: { type: String, enum: ['approved', 'pending', 'flagged', 'removed'], default: 'approved' },
      reports: {
        type: [
          {
            user: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
            reason: String,
            createdAt: { type: Date, default: Date.now },
          },
        ],
        default: [],
      },
    },
  },
  { timestamps: true }
);

routeSchema.index({ startPoint: '2dsphere' });
routeSchema.index({ country: 1, difficulty: 1 });
routeSchema.index({ 'moderation.status': 1 });

// Viewer-facing projection — matches the Piko frontend `Route` type.
routeSchema.methods.toClient = function toClient(userId) {
  const uid = userId ? userId.toString() : null;
  const votes = this.votes || [];
  const userVote = uid ? votes.find((v) => v.user && v.user.toString() === uid)?.value ?? 0 : 0;
  const voteScore = votes.reduce((sum, v) => sum + (v.value || 0), 0);
  const isSaved = uid ? (this.savedBy || []).some((u) => u.toString() === uid) : false;
  const hasGeo = this.geometry && Array.isArray(this.geometry.coordinates) && this.geometry.coordinates.length > 1;
  const hasStart = this.startPoint && Array.isArray(this.startPoint.coordinates) && this.startPoint.coordinates.length === 2;

  return {
    id: this.slug || this._id.toString(),
    _id: this._id.toString(),
    type: this.type,
    title: this.title,
    country: this.country,
    location: this.location,
    description: this.description,
    difficulty: this.difficulty,
    distanceKm: this.distanceKm,
    elevationGainM: this.elevationGainM,
    estimatedDurationMins: this.estimatedDurationMins,
    photos: this.photos || [],
    tags: this.tags || [],
    ecoScore: this.ecoScore,
    weatherScore: this.weatherScore,
    accessibilityScore: this.accessibilityScore,
    ecoImpact: this.ecoImpact,
    geometry: hasGeo ? { type: 'LineString', coordinates: this.geometry.coordinates } : undefined,
    startPoint: hasStart ? this.startPoint.coordinates : null,
    creatorName: this.creatorName,
    moderationStatus: this.moderation ? this.moderation.status : 'approved',
    isSaved,
    saveCount: (this.savedBy || []).length,
    voteScore,
    userVote,
    commentCount: (this.comments || []).length,
  };
};

module.exports = mongoose.model('Route', routeSchema);

File: dayla_flutter/lib/features/piko/data/models/route_model.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'route_model.freezed.dart';
part 'route_model.g.dart';

/// A Piko trail route, matching the backend `route.model.js` `toClient` shape.
@freezed
abstract class RouteModel with _$RouteModel {
  const factory RouteModel({
    required String id,
    @Default('curated') String type,
    @Default('') String title,
    @Default('') String country,
    @Default('') String location,
    @Default('') String description,
    @Default('moderate') String difficulty,
    @Default(0) double distanceKm,
    @Default(0) double elevationGainM,
    @Default(0) double estimatedDurationMins,
    @Default([]) List<String> photos,
    @Default([]) List<String> tags,
    @Default(75) double ecoScore,
    @Default(70) double weatherScore,
    @Default(50) double accessibilityScore,
    RouteEcoImpact? ecoImpact,
    RouteGeometry? geometry,
    List<double>? startPoint,
    @Default('Dayla') String creatorName,
    @Default('approved') String moderationStatus,
    @Default(false) bool isMine,
    @Default(false) bool isSaved,
    @Default(0) int saveCount,
    @Default(0) int voteScore,
    @Default(0) int userVote,
    @Default(0) int commentCount,
  }) = _RouteModel;

  factory RouteModel.fromJson(Map<String, dynamic> json) =>
      _$RouteModelFromJson(_normalize(json));

  static Map<String, dynamic> _normalize(Map<String, dynamic> json) {
    final copy = Map<String, dynamic>.from(json);
    copy['id'] ??= copy['_id'];
    return copy;
  }
}

@freezed
abstract class RouteEcoImpact with _$RouteEcoImpact {
  const factory RouteEcoImpact({
    @Default('') String transportMode,
    @Default(0) double co2EstimateKg,
    @Default([]) List<String> greenerAlternatives,
  }) = _RouteEcoImpact;

  factory RouteEcoImpact.fromJson(Map<String, dynamic> json) =>
      _$RouteEcoImpactFromJson(json);
}

/// GeoJSON-style LineString. Coordinates are `[lng, lat]` with an optional
/// third elevation value.
@freezed
abstract class RouteGeometry with _$RouteGeometry {
  const factory RouteGeometry({
    @Default('LineString') String type,
    @Default([]) List<List<double>> coordinates,
  }) = _RouteGeometry;

  factory RouteGeometry.fromJson(Map<String, dynamic> json) =>
      _$RouteGeometryFromJson(json);
}

@freezed
abstract class RouteCommentModel with _$RouteCommentModel {
  const factory RouteCommentModel({
    required String id,
    @Default('Anonymous') String author,
    @Default('') String content,
    String? createdAt,
  }) = _RouteCommentModel;

  factory RouteCommentModel.fromJson(Map<String, dynamic> json) =>
      _$RouteCommentModelFromJson(json);
}

/// A group plan (Dayla trip + dashboard) a route can be added to.
/// Assembled client-side from `/api/trips` + `/api/board/by-trip`.
class PikoPlan {
  const PikoPlan({
    required this.tripId,
    required this.name,
    this.subtitle,
  });

  final String tripId;
  final String name;
  final String? subtitle;
}
